src/main/java/com/thebois/utils/MatrixUtils.java

```diff
@@ -27,4 +45,108 @@ private MatrixUtils() {
         }
     }
 
+    /**
+     * Traverses the given Optional matrix using a spiral pattern to find present element
+     * within the search radius.
+     *
+     * @param matrix Matrix to be searched.
+     *
+     * @param startRow Row to start search from.
+     *
+     * @param startCol Column to start search from.
+     *
+     * @param maxSearchRadius Maximum search radius.
+     *
+     * @param maxFoundElements Maximum needed elements to be found.
+     *
+     * @param <TType> Generic type of Optional matrix.
+     *
+     * @return First found present Optional element of type TType.
+     */
+    public static <TType> Collection<Optional<TType>> matrixSpiralSearch(
+        final Optional<TType>[][] matrix,
+        final int startRow,
+        final int startCol,
+        final int maxSearchRadius,
+        final int maxFoundElements) {
+
+        // Spiral search pattern
+        final int[][] searchPattern = {
+            {0, 0}, {0, 1}, {1, 1},
+            {1, 0}, {1, -1}, {0, -1},
+            {-1, -1}, {-1, 0}, {-1, 1},
+            };
+        final Collection<Optional<TType>> foundElements = new ArrayList<>();
+
+        for (int searchRadius = 1; searchRadius <= maxSearchRadius; searchRadius++) {
+            for (final int[] ints : searchPattern) {
+                try {
+                    final int searchRow = startRow + ints[0] * searchRadius;
+                    final int searchCol = startCol + ints[1] * searchRadius;
+
+                    final Optional<TType> currentElem =
+                        matrix[searchRow][searchCol];
+
+                    if (currentElem.isPresent()) {
+                        foundElements.add(matrix[searchRow][searchCol]);
+                    }
+                }
+                catch (final ArrayIndexOutOfBoundsException exception) {
+                    // Since there is a search limit, we can let the method loop outside of matrix
+                }
+            }
+            if (foundElements.size() >= maxFoundElements) {
+                break;
+            }
+        }
+
+        // If no structure was found
+        if (foundElements.isEmpty()) {
+            foundElements.add(Optional.empty());
+        }
+        return foundElements;
+    }
+
+    /**
+     * Returns a sub matrix based on the given matrix.
+     *
+     * @param matrix   The matrix to be used to generate a sub matrix from
+     * @param startRow Row to start at
+     * @param startCol Column to start at
+     * @param endRow   Row to end at
+     * @param endCol   Column to end at
+     * @param <TType>  The generic type of the Optional matrix
+     *
+     * @return The generated sub matrix
+     *
+     * @throws IllegalArgumentException Entered indices can't be negative
+     */
+    public static <TType> Optional<TType>[][] getSubMatrix(final Optional<TType>[][] matrix,
```

Comment:
Might have missed or not gotten to that point but where is this used? And why do we need it?